### PR TITLE
Build Network Fixtures and Run Tests with Kitchen-Terraform

### PR DESCRIPTION
**Pull request by Jberlinsky**
Continuation of https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/20, incorporating https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/20#issuecomment-434447729.

Addresses https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/issues/28 and https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/issues/25 along the way.

### CODE REVIEW

File: test/integration/node_pool/controls/gcloud.rb [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    let!(:data) do
+      if subject.exit_status == 0
+        JSON.parse(subject.stdout)
+      else
+        {}
+      end
+    end
+
+    describe "default node pool" do
+      let(:default_node_pool) { data['nodePools'].select { |p| p['name'] == "default-pool" }.first }
```
Comment (aaron-lane) [inline test/integration/node_pool/controls/gcloud.rb]: Possible target for `include` / `including`.

File: test/integration/node_pool/controls/gcloud.rb [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+    describe "node pools" do
+      let(:node_pools) { data['nodePools'].reject { |p| p['name'] == "default-pool" } }
+
+      it "has 2" do
+        expect(node_pools.count).to eq 2
+      end
+
+      describe "pool-01" do
+        let(:node_pool) { node_pools.select { |p| p['name'] == "pool-01" }.first }
```
Comment (aaron-lane) [inline test/integration/node_pool/controls/gcloud.rb]: Possible target for `include` / `including`.

File: test/integration/node_pool/controls/gcloud.rb [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            "all-node-example",
+            "pool-01-example",
+            "gke-node-pool-cluster",
+            "gke-node-pool-cluster-pool-01",
+          ])
+        end
+      end
+
+      describe "pool-02" do
+        let(:node_pool) { node_pools.select { |p| p['name'] == "pool-02" }.first }
```
Comment (aaron-lane) [inline test/integration/node_pool/controls/gcloud.rb]: Possible target for `include` / `including`.

File: test/integration/node_pool/controls/kubectl.rb [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        },
+        auth_options: {
+          bearer_token: Base64.decode64(client_token),
+        },
+      )
+    end
+
+    describe "nodes" do
+      let(:all_nodes) { client.get_nodes }
+      let(:taints) { nodes.first.spec.taints.map { |t| t.to_h.select { |k, v| [:effect, :key, :value].include?(k.to_sym) } } }
```
Comment (aaron-lane) [inline test/integration/node_pool/controls/kubectl.rb]: Possible target for `include` / `including`.

File: test/integration/node_pool/controls/kubectl.rb [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+      describe "pool-01" do
+        let(:nodes) do
+          all_nodes.select { |n| n.metadata.labels.node_pool == "pool-01" }
```
Comment (aaron-lane) [inline test/integration/node_pool/controls/kubectl.rb]: Possible target for `include` / `including`.

File: test/integration/node_pool/controls/kubectl.rb [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+              key: "pool-01-example",
+              value: "true",
+            },
+          ])
+        end
+      end
+
+      describe "pool-02" do
+        let(:nodes) do
+          all_nodes.select { |n| n.metadata.labels.node_pool == "pool-02" }
```
Comment (aaron-lane) [inline test/integration/node_pool/controls/kubectl.rb]: Possible target for `include` / `including`.

File: test/integration/simple_regional/controls/gcloud.rb [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+          },
+          "networkPolicyConfig" => {
+            "disabled" => true,
+          },
+        })
+      end
+    end
+
+    describe "default node pool" do
+      let(:default_node_pool) { data['nodePools'].select { |p| p['name'] == "default-pool" }.first }
```
Comment (aaron-lane) [inline test/integration/simple_regional/controls/gcloud.rb]: Possible target for `include` / `including`.

File: test/integration/simple_zonal/controls/gcloud.rb [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+          },
+          "networkPolicyConfig" => {
+            "disabled" => true,
+          },
+        })
+      end
+    end
+
+    describe "default node pool" do
+      let(:default_node_pool) { data['nodePools'].select { |p| p['name'] == "default-pool" }.first }
```
Comment (aaron-lane) [inline test/integration/simple_zonal/controls/gcloud.rb]: Possible target for `include` / `including`.

File: test/integration/node_pool/controls/gcloud.rb [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        expect(default_node_pool['initialNodeCount']).to eq nil
+      end
+
+      it "does not have autoscaling enabled" do
+        expect(default_node_pool['autoscaling']).to eq nil
+      end
+    end
+
+    describe "node pools" do
+      let(:node_pools) { data['nodePools'].reject { |p| p['name'] == "default-pool" } }
```
Comment (aaron-lane) [inline test/integration/node_pool/controls/gcloud.rb]: Possible target for `include` / `including`.

File: test/integration/simple_regional/controls/gcloud.rb [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+        expect(default_node_pool['initialNodeCount']).to eq nil
+      end
+
+      it "does not have autoscaling enabled" do
+        expect(default_node_pool['autoscaling']).to eq nil
+      end
+    end
+
+    describe "node pool" do
+      let(:node_pool) { data['nodePools'].reject { |p| p['name'] == "default-pool" }.first }
```
Comment (aaron-lane) [inline test/integration/simple_regional/controls/gcloud.rb]: Possible target for `include` / `including`.

File: test/integration/simple_zonal/controls/gcloud.rb [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        expect(default_node_pool['initialNodeCount']).to eq nil
+      end
+
+      it "does not have autoscaling enabled" do
+        expect(default_node_pool['autoscaling']).to eq nil
+      end
+    end
+
+    describe "node pool" do
+      let(:node_pool) { data['nodePools'].reject { |p| p['name'] == "default-pool" }.first }
```
Comment (aaron-lane) [inline test/integration/simple_zonal/controls/gcloud.rb]: Possible target for `include` / `including`.

File: test/integration/node_pool/controls/gcloud.rb [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+      it "has 2" do
+        expect(node_pools.count).to eq 2
+      end
+
+      describe "pool-01" do
+        let(:node_pool) { node_pools.select { |p| p['name'] == "pool-01" }.first }
+
+        it "exists" do
+          expect(node_pool).not_to be_nil
+          expect(node_pool['name']).to eq "pool-01"
```
Comment (aaron-lane) [inline test/integration/node_pool/controls/gcloud.rb]: Target for new example.

File: test/integration/node_pool/controls/gcloud.rb [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+          ])
+        end
+      end
+
+      describe "pool-02" do
+        let(:node_pool) { node_pools.select { |p| p['name'] == "pool-02" }.first }
+
+        it "exists" do
+          expect(node_pool).not_to be_nil
+          expect(node_pool['name']).to eq "pool-02"
```
Comment (aaron-lane) [inline test/integration/node_pool/controls/gcloud.rb]: Target for new example.

File: test/integration/simple_zonal/controls/gcloud.rb [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+    end
+
+    describe "cluster" do
+      it "is running" do
+        expect(data['status']).to eq 'RUNNING'
+      end
+
+      it "is zonal" do
+        expect(data['zone']).to eq location
+        expect(data['zone']).not_to eq region
```
Comment (aaron-lane) [inline test/integration/simple_zonal/controls/gcloud.rb]: Target for new example.

File: Makefile [hunk 15]
```diff
@@ -63,20 +74,79 @@ check_headers:
...
+# Integration tests
+.PHONY: test_integration
+test_integration:
+	bundle install
+	bundle exec kitchen create
+	bundle exec kitchen converge
+	bundle exec kitchen converge
+	bundle exec kitchen verify
+	bundle exec kitchen destroy
+	cd test/fixtures/networks/ && terraform destroy -auto-approve && cd -
```
[the remark this answers is not in the document]
Comment (Jberlinsky) [reply Makefile]: Unfortunately not. The lifecycle hook actually runs at the end of every `test-kitchen` suite, not at the end of all of the suites. None of the lifecycles documented (https://github.com/test-kitchen/test-kitchen/blob/master/RELEASE_NOTES.md) will run solely before or after the entire set of suites.

We're able to get away with this in `pre_converge` because applying the fixtures when they already exist is an idempotent (and relatively quick) operation -- and one could argue that it makes sense that if a test case, somehow, messes with the fixture networks, it is sensible to "repair" them.

File: test/integration/node_pool/controls/gcloud.rb (hunk 1 above)
Comment (Jberlinsky) [reply test/integration/node_pool/controls/gcloud.rb]: Is something like this what you're getting at?

```ruby
    describe "default node pool" do
      let(:node_pools) { data['nodePools'].map(&:to_h) }
      let(:default_node_pool) { data['nodePools'].select { |p| p['name'] == "default-pool" }.first }

      it "exists" do
        expect(node_pools).to include(
          including(
            "name" => "default-pool",
          )
        )
      end

      it "has no initial node count" do
        expect(default_node_pool['initialNodeCount']).to eq nil
      end

      it "does not have autoscaling enabled" do
        expect(default_node_pool['autoscaling']).to eq nil
      end
    end
```

In this case, the principal difference between what already exists and what I've pasted above is an explicit example case that asserts there is a node pool named `default-pool`. A failure would happen either way, but the error message as currently written would admittedly require a bit more investigation to reach the conclusion that `data['nodePools'].select { |p| p['name'] == "default-pool" } == []`, and that `#first` had been called on the empty array. I'm all for providing clear error messages -- just want to make sure this is what you're getting at.

File: Makefile (hunk 15 above)
[the remark this answers is not in the document]
Comment (aaron-lane) [reply Makefile]: Ah, okay. Makes sense!

File: test/integration/node_pool/controls/gcloud.rb (hunk 1 above)
Comment (aaron-lane) [reply test/integration/node_pool/controls/gcloud.rb]: Yup, but we could go even further in this case since all three examples are really dealing with the same hash:

```ruby
describe "default node pool" do
  it "has no initial node count or autoscaling" do
    expect(node_pools).to include(
      including(
        "name" => "default-pool",
        "initialNodeCount" => nil,
        "autoscaling" => nil
      )
    )
  end
end
```

A failure of this example should indicate if the "default-pool" element is missing or if it violates one of the other constraints. It may be preferable to leave the cases as discrete examples, but even still, we can use the `include` / `including` matcher to avoid the out of band filtering of the node pools.

Comment (Jberlinsky) [reply test/integration/node_pool/controls/gcloud.rb]: Ah, that's a fair point. Thanks for clarifying!

**Comment by Jberlinsky** [comment]
@morgante I haven't seen the `kitchen converge` error you mentioned before. Could you expand on how your project was set up, and possibly provide a full log of the test run?

**Comment by Jberlinsky** [comment]
@morgante I've addressed the action items you listed above, based on our conversation on Thursday:

- Documentation has been updated to better reflect that of `project-factory`
- The modules in `examples/` are encapsulated by test cases in `test/fixtures/<example_name>`, which creates the requisite fixture network and calls the example as a module. These fixtures are can vary by test case, which will leave ample flexibility to cover shared VPC cases.
- The issue you encountered with `kitchen converge` in a project created by `project-factory` has been resolved.

Could you take another look at this when you get a moment?

File: .ruby-version [hunk 16]
```diff
@@ -0,0 +1 @@
+2.5.1
```
Comment (aaron-lane) [inline .ruby-version]: We should use 2.5.3 :smile:

File: README.md [hunk 17]
```diff
@@ -195,27 +196,56 @@ make generate_docs
...
+  2. `kitchen converge` creates the underlying resources. Run `kitchen converge <INSTANCE_NAME>` to create resources for a specific test case.
+  3. `kitchen verify` tests the created infrastructure. Run `kitchen verify <INSTANCE_NAME>` to run a specific test case.
+  4. `kitchen destroy` tears down the underlying resources created by `kitchen converge`. Run `kitchen destroy <INSTANCE_NAME>` to tear down resources for a specific test case.
+
+Alternatively, you can simply run `CREDENTIALS_FILE="credentials.json"  make test_integration_docker` to run all the test steps non-interactively.
+
+#### Test configuration
+
+Each test-kitchen instance is configured with a `variables.tfvars` file in the test fixture directory, e.g. `test/fixtures/node_pool/terraform.tfvars`. For convenience, since all of the variables are project-specific, these files have been symlinked to `test/fixtures/shared/terraform.tfvars`. Similarly, each test fixture has a `variables.tf` to define these variables, and an `outputs.tf` to facilitate providing necessary information for `inspec` to locate and query against created resources.
```
Comment (aaron-lane) [inline README.md]: This line should be split in to multiple lines.

File: test/fixtures/deploy_service/example.tf [hunk 18]
```diff
@@ -14,22 +14,15 @@
...
-  source             = "../../"
-  project_id         = "${var.project_id}"
-  name               = "shared-vpc-sample-cluster"
-  region             = "${var.region}"
-  network            = "${var.network}"
-  network_project_id = "${var.network_project_id}"
-  subnetwork         = "${var.subnetwork}"
-  ip_range_pods      = "${var.ip_range_pods}"
-  ip_range_services  = "${var.ip_range_services}"
+  project_id = "${var.project_id}"
```
[the remark this answers is not in the document]
Comment (Jberlinsky) [reply test/fixtures/deploy_service/example.tf]: Hm -- not sure how `make -s` missed this. I'll fix it up. Thanks!

File: test/fixtures/deploy_service/outputs.tf [hunk 19]
```diff
@@ -0,0 +1 @@
+../shared/outputs.tf
```
[the remark this answers is not in the document]
Comment (Jberlinsky) [reply test/fixtures/deploy_service/outputs.tf]: This is a symlink

File: test/make.sh [hunk 20]
```diff
@@ -48,7 +48,7 @@ function docker() {
 function check_terraform() {
   echo "Running terraform validate"
   #shellcheck disable=SC2156
-  find . -name "*.tf" -exec bash -c 'terraform validate --check-variables=false $(dirname "{}")' \;
+  find . -name "*.tf" -not -path "./test/fixtures/shared/*" -not -path "./test/fixtures/all_examples/*" -exec bash -c 'terraform validate --check-variables=false $(dirname "{}")' \;
```
[the remark this answers is not in the document]
Comment (Jberlinsky) [reply test/make.sh]: These contain partial configurations. The files in these directories are included in other functional configurations, which will be `terraform validate`d themselves.

File: .ruby-version (hunk 16 above)
Comment (Jberlinsky) [reply .ruby-version]: Will bump!

File: README.md (hunk 17 above)
Comment (Jberlinsky) [reply README.md]: Done!

**Comment by Jberlinsky** [comment]
@aaron-lane I've addressed your comments -- would appreciate a re-review!

File: README.md [hunk 21]
```diff
@@ -195,27 +196,58 @@ make generate_docs
...
+  2. `kitchen converge` creates the underlying resources. Run `kitchen converge <INSTANCE_NAME>` to create resources for a specific test case.
+  3. `kitchen verify` tests the created infrastructure. Run `kitchen verify <INSTANCE_NAME>` to run a specific test case.
+  4. `kitchen destroy` tears down the underlying resources created by `kitchen converge`. Run `kitchen destroy <INSTANCE_NAME>` to tear down resources for a specific test case.
+
+Alternatively, you can simply run `CREDENTIALS_FILE="credentials.json"  make test_integration_docker` to run all the test steps non-interactively.
+
+#### Test configuration
+
+Each test-kitchen instance is configured with a `variables.tfvars` file in the test fixture directory, e.g. `test/fixtures/node_pool/terraform.tfvars`.
+For convenience, since all of the variables are project-specific, these files have been symlinked to `test/fixtures/shared/terraform.tfvars`.
```
Comment (morgante) [inline README.md]: Err, I'm not seeing that?

Comment (morgante) [reply README.md]: Fixed myself in 97ed368478660871d7cf512621b2368caced4eb4.

File: cluster_zonal.tf [hunk 22]
```diff
@@ -134,3 +134,18 @@ resource "google_container_node_pool" "zonal_pools" {
 
   depends_on = ["google_container_cluster.zonal_primary"]
 }
+
+resource "null_resource" "wait_for_zonal_cluster" {
```
Comment (morgante) [inline cluster_zonal.tf]: This should probably go in its own PR.

File: scripts/kubectl_wrapper.sh [hunk 23]
```diff
@@ -39,7 +39,9 @@ mkdir "${TMPDIR}"
 
 export KUBECONFIG="${TMPDIR}/config"
 
-echo "${CA_CERTIFICATE}" | base64 --decode > "${TMPDIR}/ca_certificate"
+# shellcheck disable=SC1117
+base64 --help | grep "\--decode" && B64_ARG="--decode" || B64_ARG="-d"
```
Comment (morgante) [inline scripts/kubectl_wrapper.sh]: This seems very complex? Why do we need to do all this greping?

Anyways. This is not a testing change = separate PR please.

File: scripts/wait-for-cluster.sh [hunk 24]
```diff
@@ -13,13 +13,21 @@
 # See the License for the specific language governing permissions and
```
Comment (morgante) [inline scripts/wait-for-cluster.sh]: Back this into a new PR.

File: test/fixtures/shared/variables.tf [hunk 25]
```diff
@@ -14,23 +14,15 @@
  * limitations under the License.
  */
 
-output "name_example" {
-  description = "Cluster name"
-  value       = "${module.gke.name}"
-}
+variable "project_id" {}
```
Comment (morgante) [inline test/fixtures/shared/variables.tf]: As per Terraform Standards, all variables must have a description. Especially for ones which tests depend on...

**Comment by morgante** [comment]
Running `kitchen converge` twice causes a permadiff for networking:

```
         network:    "projects/clf-cft-testing-dev/global/networks/cft-gke-test-cwky" => "https://www.googleapis.com/compute/v1/projects/clf-cft-testing-dev/global/networks/cft-gke-test-cwky"
         subnetwork: "projects/clf-cft-testing-dev/regions/us-east4/subnetworks/cft-gke-test-cwky" => "https://www.googleapis.com/compute/v1/projects/clf-cft-testing-dev/regions/us-east4/subnetworks/cft-gke-test-cwky"
```

This is possibly an upstream issue.

File: cluster_zonal.tf [hunk 26]
```diff
@@ -24,7 +24,7 @@ resource "google_container_cluster" "zonal_primary" {
   project     = "${var.project_id}"
 
   zone             = "${var.zones[0]}"
-  additional_zones = "${slice(var.zones,1,length(var.zones))}"
+  additional_zones = ["${slice(var.zones,1,length(var.zones))}"]
```
[the remark this answers is not in the document]
Comment (Jberlinsky) [reply cluster_zonal.tf]: Extracted to https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/44

File: test/fixtures/shared/variables.tf [hunk 27]
```diff
@@ -14,15 +14,24 @@
...
  */
 
-variable "project_id" {}
+variable "project_id" {
+  description = "The GCP project to use for integration tests"
+}
 
-variable "credentials_path_relative" {}
+variable "credentials_path_relative" {
+  description = ...
```
Comment (morgante) [inline test/fixtures/shared/variables.tf]: This is not a proper description.

Comment (Jberlinsky) [reply test/fixtures/shared/variables.tf]: Yep, sorry -- accidentally pushed an interstitial state. Fixed.

File: scripts/kubectl_wrapper.sh (hunk 23 above)
Comment (aaron-lane) [reply scripts/kubectl_wrapper.sh]: This identifies which flag is used to "decode"; the interface of `base64` differs across platforms.

File: scripts/wait-for-cluster.sh (hunk 24 above)
Comment (Jberlinsky) [reply scripts/wait-for-cluster.sh]: See https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/46

File: scripts/kubectl_wrapper.sh (hunk 23 above)
Comment (Jberlinsky) [reply scripts/kubectl_wrapper.sh]: I'd argue that this is indeed a testing change -- in moving from running these tests on the local host (the current state of `master`) to `docker`, it was necessary to change the interface of `base64` as Aaron mentioned above depending on where the test is being run.

File: Makefile [hunk 28]
```diff
@@ -15,6 +15,17 @@
 # Make will use bash instead of sh
 SHELL := /usr/bin/env bash
 
+# Docker build config variables
```
[the remark this answers is not in the document]
Comment (Jberlinsky) [reply Makefile]: I took inspiration from `project-factory` here: https://github.com/terraform-google-modules/terraform-google-project-factory/blob/master/Makefile#L18

File: cluster_zonal.tf (hunk 22 above)
Comment (Jberlinsky) [reply cluster_zonal.tf]: See https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/46

**Comment by Jberlinsky** [comment]
@morgante At a quick glance, the permadiff does indeed look to be upstream. I'm digging into it now.

**Comment by Jberlinsky** [comment]
It does indeed look like the permadiff is upstream -- all the way up to `terraform` core. See:

- https://github.com/terraform-providers/terraform-provider-google/issues/1566
- https://github.com/hashicorp/terraform/issues/18209
